atspi-common/src/object_match.rs

```diff
@@ -0,0 +1,281 @@
+	phantom: std::marker::PhantomData<()>,
+}
+
+// !!! WARNING !!! :
+//
+// State and Role are defined as u32 in Accessible.xml but as i32 in Collection.xml
+//
+// The signature on StateSet is defined inconsistently in the XMLs
+// Accessible.xml: GetState type="au"
+// Collection.xml: GetMatches argument name="rule" type="(aiia{ss}iaiiasib)"
+// The latter starts with ai, which is a signature for an array of signed ints i32.
+//
+// https://gitlab.gnome.org/federico/at-spi2-core/-/commit/4885efedeef71e0df8210622771a0b1bb10e194d
+impl Type for ObjectMatchRule {
+	fn signature() -> Signature<'static> {
+		Signature::from_str_unchecked("(aiia{ss}iaiiasib)")
+	}
+}
+
+impl ObjectMatchRule {
+	/// Create a new `MatchRuleBuilder`
+	#[must_use]
+	pub fn builder() -> ObjectMatchRuleBuilder {
+		ObjectMatchRuleBuilder::default()
+	}
+}
+
+/// The 'builder' type for `MatchRule`.  
+/// Use its methods to set match criteria.
+#[derive(Debug, Clone, Default)]
+pub struct ObjectMatchRuleBuilder {
+	states: StateSet,
+	states_mt: MatchType,
+	attr: HashMap<String, String>,
+	attr_mt: MatchType,
+	roles: Vec<Role>,
+	roles_mt: MatchType,
+	ifaces: InterfaceSet,
+	ifaces_mt: MatchType,
+	invert: bool,
+}
+
+impl ObjectMatchRuleBuilder {
+	/// Insert a `StateSet` to the builder
+	#[must_use]
+	pub fn states(mut self, state_set: StateSet, mt: MatchType) -> Self {
+		self.states = state_set;
+		self.states_mt = mt;
+		self
+	}
+
+	/// Insert a map of attributes
+	#[must_use]
+	pub fn attributes(mut self, attributes: HashMap<String, String>, mt: MatchType) -> Self {
+		self.attr = attributes;
+		self.attr_mt = mt;
+		self
+	}
+
+	/// Insert a slice of `Role`s
+	#[must_use]
+	pub fn roles(mut self, roles: &[Role], mt: MatchType) -> Self {
+		self.roles = roles.into();
+		self.roles_mt = mt;
+		self
+	}
+
+	/// Insert a slice of `Interface`s
+	#[must_use]
+	pub fn interfaces(mut self, interfaces: &[Interface], mt: MatchType) -> Self {
+		self.ifaces = interfaces.iter().copied().collect::<InterfaceSet>();
+		self.ifaces_mt = mt;
+		self
+	}
+
+	/// Sets the inversion of the `MatchRule`, defaults to `false`, no inversion.
+	#[must_use]
+	pub fn invert(mut self, invert: bool) -> Self {
+		self.invert = invert;
+		self
+	}
+
+	/// Builds the [`ObjectMatchRule`]
+	///
+	/// [`ObjectMatchRule``]: crate::object_match::ObjectMatchRule
+	#[must_use]
+	pub fn build(self) -> ObjectMatchRule {
+		ObjectMatchRule {
+			states: self.states,
+			states_mt: self.states_mt,
+			attr: self.attr,
+			attr_mt: self.attr_mt,
+			roles: self.roles,
+			roles_mt: self.roles_mt,
+			ifaces: self.ifaces,
+			ifaces_mt: self.ifaces_mt,
+			invert: self.invert,
+			phantom: PhantomData,
+		}
+	}
+}
+
+/// Enumeration used by [`ObjectMatchRule`] to specify how to interpret [`ObjectRef`] objects.
+///
+/// [`ObjectRef`]: crate::object_ref::ObjectRef
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, Type, Default)]
+#[repr(i32)]
+pub enum MatchType {
+	#[default]
```

Comment:
Unsure if this should be the default. In my mind, a default `ObjectMatch` will match on _all_ objects. Perhaps this is sort of a mistaken idea, but that's where it comes from.